docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+it will create a ``Notification`` object and attach it to the particular ``Build``.
+Note the director does not have access to the DB, so it will need to create/associate the object via an API endpoint/Celery task.
+
+Example of how the exception ``BuildCancelled`` creates an error ``Notification``:
+
+
+.. code-block:: python
+
+    class UpdateDocsTask(...):
+        def on_failure(self):
+            self.data.api_client.build(self.data.build["id"]).notifications.post(
+                {
+                    "slug": "cancelled-by-user",
+                    # Override default fields if required
+                    "type": WARNING,
+                }
+            )
+
+
+
+Attach non-error ``Notification`` during the build process
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+During the build, we will be able attach non-error notifications with the following pattern:
+
+- check something in particular (e.g. using a deprecated key in ``readthedocs.yaml``)
+- create a non-error ``Notification`` and attach it to the particular ``Build`` object
+
+.. code-block:: python
+
+    class DockerBuildEnvironment(...):
+        def check_deprecated_os_image(self):
+            if self.config.build.os == "ubuntu-18.04":
+                self.api_client.build(self.data.build["id"]).notifications.post(
+                    {
+                        "slug": "os-ubuntu-18.04-deprecated",
+                    }
+                )
+
+
+
+Show a ``Notification`` under the user's bell icon
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If we want to show a notification on a user's profile,
+we can create the notification as follows,
+maybe from a simple script ran in the Django shell's console
+after publishing a blog post:
+
+
+.. code-block:: python
+
+    users_to_show_notification = User.objects.filter(...)
+
+    for user in users_to_show_notification:
+        Notification.objects.create(
+            slug="blog-post-beta-addons",
+            dismissable=True,
+            bell=True,
+            attached_to=User,
+            attached_to_id=user.id,
+        )
+
+
+Remove notification on status change
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When we show a notification for an unpaid subscription,
+we want to remove it once the user has updated and paid the subscription.
+We can do this with the following code:
+
+
+.. code-block:: python
+
+    @handler("customer.subscription.updated", "customer.subscription.deleted")
+    def subscription_updated_event(event):
+        if subscription.status == ACTIVE:
+            organization = Organization.objects.get(slug="read-the-docs")
+
+            Notification.objects.filter(
+                slug="subscription-update-your-cc-details",
+                state__in=[UNREAD, READ],
+                attached_to=Organization,
+                attached_to_id=organization.id,
+            ).update(state=DISMISSED)
```

Comment:
Do we maybe want to distinguish an automatically dismissed notification with something like `state=CANCELLED`?

Reply:
Sounds good. Done in e3773728